src/metrics/include/metrics/v2/event.hpp

```diff
@@ -247,70 +285,269 @@ class Collector {
+
+// Manages a thread of grpc client execution .
+template <typename ClockSource, typename StreamInterface>
+class GrpcThread {
+public:
+    GrpcThread() : _thread{&GrpcThread::run, this} {}
+
+    typedef EventStream<ClockSource, StreamInterface>* stream_ptr;
+
+    void registerStream(stream_ptr stream) {
+        const std::lock_guard<std::mutex> lock(_streams_mutex);
+        streams.insert(stream);
+    }
+
+    void closeStream(stream_ptr stream) {
+        const std::lock_guard<std::mutex> lock(_streams_mutex);
+        streams.erase(stream);
+    }
+
+    ~GrpcThread() {
+        done = true;
+        _thread.join();
+    }
+
+private:
+    void run() {
+        while (!done || !streams.empty()) {
+            reap_actors();
+            std::this_thread::sleep_for(std::chrono::milliseconds(GRPC_THREAD_SLEEP_MS));
+        }
+    }
+
+    void reap_actors() {
+        bool still_reaping = false;
+        do {
+            still_reaping = false;
+            const std::lock_guard<std::mutex> lock(_streams_mutex);
+            for (auto stream : streams) {
+                if (stream->sendOne())
+                    still_reaping = true;
+            }
+        } while (still_reaping);
+    }
+
+    std::atomic<bool> done = false;
+    std::thread _thread;
+    std::set<stream_ptr> streams;
+    std::mutex _streams_mutex;
+};
+
+// Manages all the grpc threads. Divides the workload evenly between them.
+template <typename ClockSource, typename StreamInterface>
+class GrpcClient {
+public:
+    GrpcClient() : _threads(CLIENT_THREADS) {}
+
+    // EventStream users manage their own memory and should deregister themselves before
+    // destructing, so we (carefully) use naked pointers.
+    typedef EventStream<ClockSource, StreamInterface>* stream_ptr;
+
+    void registerStream(stream_ptr stream) {
+        _threads[latest_thread++ % _threads.size()].registerStream(stream);
+    }
+
+    void closeStream(stream_ptr stream) {
+        for (int i = 0; i < _threads.size(); i++) {
+            _threads[i].closeStream(stream);
+        }
+    }
+
+private:
+    std::vector<GrpcThread<ClockSource, StreamInterface>> _threads;
+    std::atomic<size_t> latest_thread = 0;
+};
+
+template <typename ClockSource>
+struct MetricsArgs {
+    MetricsArgs(const typename ClockSource::time_point& finish,
+                OperationEventT<ClockSource> event,
+                size_t workerCount)
+        : finish(finish), event(std::move(event)), workerCount(workerCount) {}
+    typename ClockSource::time_point finish;
+    OperationEventT<ClockSource> event;
+    size_t workerCount;
+};
+
+// Efficient buffer that should rarely be blocked on insertions.
+template <typename ClockSource>
+class MetricsBuffer {
+public:
+    explicit MetricsBuffer(size_t size, const std::string& name)
+        : size{size},
+          name{name},
+          _loading(new std::vector<MetricsArgs<ClockSource>>()),
+          _draining(new std::vector<MetricsArgs<ClockSource>>()) {
```

Comment:
I think `make_unique` is generally less surprising than `new` inside paren initializers